src/EndpointRegistry.sol

```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: Apache 2
+pragma solidity >=0.6.12 <0.9.0;
+
+/// @dev This contract is responsible for handling the registration of Endpoints.
+abstract contract EndpointRegistry {
+    constructor() {
+        _checkEndpointsInvariants();
+    }
+
+    /// @dev Information about registered endpoints.
+    struct EndpointInfo {
+        // whether this endpoint is registered
+        bool registered;
+        // whether this endpoint is enabled
+        bool enabled;
+        uint8 index;
+    }
+
+    /// @dev Bitmap encoding the enabled endpoints.
+    /// invariant: forall (i: uint8), enabledEndpointBitmap & i == 1 <=> endpointInfos[i].enabled
+    struct _EnabledEndpointBitmap {
+        uint64 bitmap;
+    }
+
+    /// @dev Total number of registered endpoints. This number can only increase.
```

Comment:
This comment is not true since we can remove endpoints?

Reply:
this is "registered" but not necessarily "enabled" (the latter of which can go down indeed). we keep the disabled ones to ensure their index is not reused